pkgs/by-name/wa/warp-terminal/update.sh

```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -I nixpkgs=./. -i bash -p curl jq common-updater-scripts nix-prefetch
+#shellcheck shell=bash
+
+set -eu -o pipefail
+
+dirname="$(dirname "$0")"
+
+# it's fine to use just the linux URL because the package also assumes the versioning is the same between darwin and linux
+version=$(curl -Ls -o /dev/null -w %{url_effective} https://app.warp.dev/download\?package\=pacman 2>&1 | grep -oP -m 1 "(?<=v)[\d.\w]+" | head -1)
+old_version=$(nix-instantiate --eval -A warp-terminal.version | jq -e -r)
+
+echo $version
+
+if [[ $version == "$old_version" ]]; then
+    echo "New version same as old version, nothing to do." >&2
+    exit 0
+fi
+
+# manually updating hash because update-source-version's --system flag expects an explicit platform + architecture, not just a generic platform
+macOS_url="https://releases.warp.dev/stable/v$version/Warp.dmg";
+macOS_hash=$(nix-prefetch-url --type sha256 $macOS_url)
+
+# this monstrosity is needed to find the second appearance of [ hash = "..."; ] and replace it with the correct hash
+sed -i '/hash = ".*";/{G;s/\nX\{1\}//;tend;x;s/^/X/;x;P;d};p;d;:end;s/hash = ".*"/hash = "'"$macOS_hash"'"/;:a;n;ba' "$dirname/package.nix"
```

Comment:
you can put the hash and version in a json file and have your update script read / write the json file, and have the derivation import the json.

see:
https://github.com/NixOS/nixpkgs/blob/268690ff59e25a06a7c782900144df92ed545584/pkgs/servers/invidious/default.nix


```json
{
	"darwin" : {
		"hash" : "sha256-tFtoD8URMFfJ3HRkyKStuDStFkoRIV97y9kV4pbDPro=",
		"version" : "0.2024.02.20.08.01.stable_01"
	},
	"linux" : {
		"hash" : "sha256-L8alnqSE4crrDozRfPaAAMkLc+5+8d9XBKd5ddsxmD0=",
		"version" : "0.2024.02.20.08.01.stable_01"
	}
}
```

then in nix
```nix
let
pname = "warp-terminal";
versions = lib.importJSON ./versions.json;

linux = stdenv.mkDerivation (finalAttrs:  {
  inherit pname meta;
  inherit (versions.linux) version;
  src = fetchurl {
    inherit (versions.linux) hash;
    url = "https://releases.warp.dev/stable/v${finalAttrs.version}/warp-terminal-v${finalAttrs.version}-1-x86_64.pkg.tar.zst";
  };
[...]
```

Reply:
Thanks for the writeup! I really appreciate the explanation instead of just changing it. I have the fixed up version that should be ready.